src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionHelpers.cs

```diff
@@ -4,58 +4,203 @@
+            {
+                foreach (var suggestedAction in set.Actions)
+                {
+                    // Filter out code actions with options since they'll show dialogs and we can't remote the UI and the options.
+                    if (suggestedAction.CodeAction is CodeActionWithOptions)
+                    {
+                        continue;
+                    }
+
+                    codeActions.Add(GetNestedActionsFromActionSet(suggestedAction));
+                }
+            }
+
+            return codeActions.ToImmutable();
+        }
 
-            var codeActions = codeFixCollections.SelectMany(c => c.Fixes.Select(f => f.Action)).Concat(
-                    codeRefactorings.SelectMany(r => r.CodeActions.Select(ca => ca.action)));
+        /// <summary>
+        /// Generates a code action with its nested actions properly set.
+        /// </summary>
+        private static CodeAction GetNestedActionsFromActionSet(IUnifiedSuggestedAction suggestedAction)
+        {
+            var codeAction = suggestedAction.CodeAction;
+            if (!(suggestedAction is UnifiedSuggestedActionWithNestedActions suggestedActionWithNestedActions))
+            {
+                return codeAction;
+            }
+
+            using var _ = ArrayBuilder<CodeAction>.GetInstance(out var nestedActions);
+            foreach (var actionSet in suggestedActionWithNestedActions.NestedActionSets)
+            {
+                foreach (var action in actionSet.Actions)
+                {
+                    nestedActions.Add(GetNestedActionsFromActionSet(action));
+                }
+            }
 
-            return codeActions;
+            return new CodeActionWithNestedActions(
+                codeAction.Title, nestedActions.ToImmutable(), codeAction.IsInlinable, codeAction.Priority);
         }
 
-        public static async Task<(ImmutableArray<CodeFixCollection>, ImmutableArray<CodeRefactoring>)> GetCodeFixesAndRefactoringsAsync(
+        private static async Task<ImmutableArray<UnifiedSuggestedActionSet>?> GetActionSetsAsync(
             Document document,
             ICodeFixService codeFixService,
             ICodeRefactoringService codeRefactoringService,
+            IThreadingContext threadingContext,
             LSP.Range selection,
             CancellationToken cancellationToken)
         {
             var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
             var textSpan = ProtocolConversions.RangeToTextSpan(selection, text);
 
-            var codeFixCollectionsTask = codeFixService.GetFixesAsync(document, textSpan, includeSuppressionFixes: true, cancellationToken);
-            var codeRefactoringsTask = codeRefactoringService.GetRefactoringsAsync(document, textSpan, cancellationToken);
+            // The logic to filter code actions requires the UI thread
+            await threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
switching to the main thread is a potential deadlock risk across the streamjsonrpc boundary.  Is there a tracking issue to remove the UI thread dependency?

Reply:
Opened #46431 to track.

Reply:
note: this is likely pretty important.  i think we only have a couple of code-actions taht have the UI thread dep.  we likely should have them expose that in some way (perhaps an interface they implement).  And then just filter those out from the LSP case